### PR TITLE
bring hip graph back

### PR DESCRIPTION
**Pull request by nimlgen**
Bring hip graph back. Now 2.5ms cpu time (gpt2) on the og tiny.

### CODE REVIEW

**Comment by geohot** [comment]
How many of these 44 lines can be shared with Metal ICB?

File: tinygrad/runtime/ops_hip.py [hunk 1]
```diff
@@ -48,6 +49,19 @@ def compile_hip(prg) -> bytes:
   hip.hiprtcCompileProgram(prog, [f'--offload-arch={hip.hipGetDeviceProperties(HIP.default_device).gcnArchName}'])
   return hip.hiprtcGetCode(prog)
 
+def time_exection(cb, enable=False):
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:52]: Typo in function name

File: tinygrad/ops.py [hunk 2]
```diff
@@ -164,6 +165,13 @@ def __init__(self, jit_cache: List[JitItem], input_rawbuffers: Dict[Union[int, s
         if a in [v for v in input_rawbuffers.values()]:
           self.input_replace[(j,i)] = [k for k,v in input_rawbuffers.items() if v == a][0]
     assert set(self.input_replace.values()) == set(input_rawbuffers.keys()), "some input tensors not found"
+
+    # Record info about updatable jit entries.
+    def dim_has_vars(dim): return dim and any(isinstance(x, Node) for x in dim)
+    self.updatable_launch_dims: List[int] = [j for j,ji in enumerate(jit_cache) if isinstance(ji.prg, CompiledASTRunner) and (dim_has_vars(ji.prg.global_size) or dim_has_vars(ji.prg.local_size))]
+    self.updatable_var_vals: List[int] = [j for j,ji in enumerate(jit_cache) if ji.prg.vars]
+    self.updatable_inputs: List[int] = [x[0] for x in self.input_replace.keys()]
```
Comment (geohot) [inline tinygrad/ops.py:173]: I'm confused about this last one. The other two should have names that reflect that they are kernel indexes

File: tinygrad/ops.py [hunk 3]
```diff
@@ -164,6 +165,13 @@ def __init__(self, jit_cache: List[JitItem], input_rawbuffers: Dict[Union[int, s
         if a in [v for v in input_rawbuffers.values()]:
           self.input_replace[(j,i)] = [k for k,v in input_rawbuffers.items() if v == a][0]
     assert set(self.input_replace.values()) == set(input_rawbuffers.keys()), "some input tensors not found"
+
+    # Record info about updatable jit entries.
+    def dim_has_vars(dim): return dim and any(isinstance(x, Node) for x in dim)
```
Comment (geohot) [inline tinygrad/ops.py:170]: Use all_int from helpers

File: tinygrad/ops.py [hunk 4]
```diff
@@ -164,6 +165,13 @@ def __init__(self, jit_cache: List[JitItem], input_rawbuffers: List[RawBuffer],
         if a in input_rawbuffers:
           self.input_replace[(j,i)] = input_rawbuffers.index(a)
     assert len(set(self.input_replace.values())) == len(input_rawbuffers), "some input tensors not found"
+
+    # Record indices of jit items components of which could be updated each run.
+    def dim_has_vars(dim): return dim and not all_int(dim)
```
Comment (geohot) [inline tinygrad/ops.py:170]: Isn't just checking not all_int enough?

File: tinygrad/runtime/ops_hip.py [hunk 5]
```diff
@@ -63,22 +77,62 @@ def __init__(self, name:str, prg:bytes):
...
-      hip.hipEventDestroy(start)
-      hip.hipEventDestroy(end)
-      return ret
+    return time_execution(lambda: hip.hipModuleLaunchKernel(self.prgs[args[0]._device], global_size[0], global_size[1], global_size[2], local_size[0], local_size[1], local_size[2], 0, 0, struct), enable=wait)
 
   def __del__(self):
     for module in self.modules: hip.hipModuleUnload(module)
 
+class HIPGraph(BatchExecutor):
+  UpdateNode = collections.namedtuple("UpdateNode", ["node", "params"])
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:87]: Not sure this being a namedtuple gets us anything, it's very local

File: tinygrad/runtime/ops_hip.py [hunk 6]
```diff
@@ -63,22 +77,62 @@ def __init__(self, name:str, prg:bytes):
...
+
+    self.instance = hip.hipGraphInstantiate(self.graph)
+    self.clear_jit_inputs()
+
+  def __del__(self):
+    if hasattr(self, 'instance'): hip.hipGraphExecDestroy(self.instance)
+    if hasattr(self, 'graph'): hip.hipGraphDestroy(self.graph)
+
+  def __call__(self, input_rawbuffers: List[RawBuffer], var_vals: Dict[Variable, int], wait=False):
+    if not hasattr(self, 'instance'): return super().__call__(input_rawbuffers, var_vals, wait)
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:118]: How does METAL deal with this...looks hacky

Comment (nimlgen) [reply tinygrad/runtime/ops_hip.py:118]: This is a case for world.py and multigpus. We can't use graphs for jit_cache which contains memory operations and not only `HIPProgram`s. Now it just switches to basic batch executor. In the future we can encode memory operations into the graph as well and events for synchronisation.

Comment (geohot) [reply tinygrad/runtime/ops_hip.py:118]: My concern is not that it switches, it's how this is being done with these `hasattr`

File: tinygrad/ops.py (hunk 4 above)
Comment (nimlgen) [reply tinygrad/ops.py:170]: Backends without locals could have None here, like LLVM

File: tinygrad/runtime/ops_hip.py [hunk 7]
```diff
@@ -63,22 +77,55 @@ def __init__(self, name:str, prg:bytes):
...
+    # Update cached params structs with the new values.
+    for (j,i),input_idx in self.input_replace.items():
+      hip.setKernelNodeParam(self.updatable_nodes[j][1], input_rawbuffers[input_idx], i)
+    for j in self.jc_idxs_with_updatable_launch_dims:
+      hip.setKernelNodeLaunchDims(self.updatable_nodes[j][1], cast(CompiledASTRunner, self.jit_cache[j].prg).launch_dims(var_vals))
+    for j in self.jc_idxs_with_updatable_var_vals:
+      hip.setKernelNodeParams(self.updatable_nodes[j][1], [var_vals[x] for x in self.jit_cache[j].prg.vars], list(range(len(self.jit_cache[j].rawbufs), len(self.jit_cache[j].rawbufs) + len(self.jit_cache[j].prg.vars))))
+
+    # Update graph nodes with the updated structs.
+    for j in self.updatable_nodes:
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:122]: `for node, params: hip.hipGraphExecKernelNodeSetParams(self.instance, node, params)`

**Comment by geohot** [comment]
I'm working on a refactor that should make this easier

**Comment by geohot** [comment]
With #2405 you can raise GraphException and it will fall back to the default behavior. Also, check out how it's not a BatchExecutor anymore, potentially allowing multiple Graphs in a JIT.

File: tinygrad/runtime/ops_hip.py [hunk 8]
```diff
@@ -63,20 +78,61 @@ def __init__(self, name:str, prg:bytes):
...
+      params = hip.buildKernelNodeParams(*ji.rawbufs, *[var_vals[x] for x in prg.vars], func=prg.clprg.prgs[ji.rawbufs[0]._device], grid=global_size, block=local_size)
+      graph_node = hip.hipGraphAddKernelNode(self.graph, [graph_node] if graph_node else [], params)
+
+      if j in self.jc_idxs_with_updatable_launch_dims or j in self.jc_idxs_with_updatable_var_vals or j in self.jc_idxs_with_updatable_rawbufs:
+        self.updatable_nodes[j] = (graph_node, params)
+
+    self.instance = hip.hipGraphInstantiate(self.graph)
+
+  def __del__(self):
+    if hasattr(self, 'instance'): hip.hipGraphExecDestroy(self.instance)
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:117]: Do you need these hasattr?

File: tinygrad/runtime/ops_hip.py (hunk 1 above)
Comment (geohot) [reply tinygrad/runtime/ops_hip.py:53]: This got reverted

File: extra/hip_wrapper.py [hunk 9]
```diff
@@ -200,21 +200,23 @@ def getStructTypeForArgs(*args):
...
-    if updated_args is not None:
-      for i in updated_args:
-        setattr(struct, f'field{i}', (args[i] if args[i].__class__ is int else args[i]._buf))
-    else:
-      for i,d in enumerate(args):
-        setattr(struct, f'field{i}', (d if d.__class__ is int else d._buf))
+  def setKernelNodeParam(npwrapper:kernelNodeParamsWrapper, arg, id):
+    setattr(npwrapper.context[1], f'field{id}', (arg if arg.__class__ is int else arg._buf))
+
+  def setKernelNodeLaunchDims(npwrapper:kernelNodeParamsWrapper, launch_dims):
```
Comment (geohot) [inline extra/hip_wrapper.py:206]: `def setKernelNodeLaunchDims(npwrapper:kernelNodeParamsWrapper, grid, block)` and `*` it when you call

File: extra/hip_wrapper.py [hunk 10]
```diff
@@ -200,21 +200,23 @@ def getStructTypeForArgs(*args):
...
 
-  def updateKernelNodeParams(npwrapper:kernelNodeParamsWrapper, *args, grid=(1,1,1), block=(1,1,1), updated_args=None):
-    _, struct, _ = npwrapper.context
-    if updated_args is not None:
-      for i in updated_args:
-        setattr(struct, f'field{i}', (args[i] if args[i].__class__ is int else args[i]._buf))
-    else:
-      for i,d in enumerate(args):
-        setattr(struct, f'field{i}', (d if d.__class__ is int else d._buf))
+  def setKernelNodeParam(npwrapper:kernelNodeParamsWrapper, arg, id):
```
Comment (geohot) [inline extra/hip_wrapper.py:203]: Actually should these helpers be here at all? Pushing complexity to extra is bad...

Comment (nimlgen) [reply extra/hip_wrapper.py]: I have removed everything tinygrad-specific from the wrapper (such as ._buf). The wrapper still contains helper functions, but I believe these are more closely related to HIP. Should I pull out more?

**Comment by geohot** [comment]
Cool, 11.5 ms -> 2.5 ms matches what I'm seeing on GPT-2 also!

**Pull request merged**